052-record-reconciled-versions.md

```diff
@@ -0,0 +1,186 @@
+Though this works, from a user standpoint it is far from ideal as it assumes a level of user knowledge about the cluster that may not be appropriate for all users.
+
+## Motivation
+
+The reason this proposal should be considered is it will allow a user to rely on the `Kafka` custom resource as their single source of truth, instead of watching/querying the individual kafka/zookeeper pods.
+This proposal lays out all the information we need and at what points in the reconcile. This should all be available from within the Reconcile loop, so orchestrating and figuring out if a component has been fully upgraded can and should be propogated into the Custom Resource via metadata.
+
+## Proposal
+
+The following is a mechanism to record reconcilliation information into a Custom Resource.
+This proposal will cover the `Kafka` and `StrimziPodSet` custom resources, but could be extended to include other Custom Resources if there would be value in it, but this proposal was designed to cover the Kafka Custom Resource as this is the core component of Strimzi, and its reliance on `StrimziPodSet` means we should implement it for this CR type as well.
+
+This proposal assumes that all key value pairs mentioned will be writen to the `metadata.annotations` field [docs here](https://kubernetes.io/docs/concepts/overview/working-with-objects/annotations/).
+But there is an argument for putting them in the [metadata.labels field](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/), but given Strimzi's heavy use of the labels field, in order to not confuse a user who might think they had accidentally set the field themselves, this proposal has chosen to use annotations.
+
+
+The following are the two proposed new annotations:
+
+The `strimzi.io/reconciled` annotation is patched at the end of a **successful** reconcile, and signals to a user that the operator at version `X` was able to reconcile to successful completion.
+This would be the annotation a user or System Test would query against to check whether upgrade was complete.
+```
+strimzi.io/reconciled: '0.37.0'
+```
+
+The `strimzi.io/reconciling` annotation is patched to the CR at the start of reconcile and signals to a user that the operator is attempting to reconcile the CR at version Y.
+```
+strimzi.io/reconciling: '0.37.0'
+```
+This annotation is required for the following reasons:
+- It signals to a user know that a reconcile on version Y has started, so that the operator is functional without checking the logs of the operator.
+- The Reconciler should update some value on reconcilliation startup without overriding the `strimzi.io/reconciled` annotation, the reason for this is, if for any reason the operator errors during reconcile it would incorrectly update `reconciled` to Y, despite not finishing to successful completion. So having a two stage commit mechanism here would give much more information to a user regarding whether it started reconcile versus whether it finished.
```

Comment:
```suggestion
- The operator updates the `strimzi.io/reconciling` annotation value on reconciliation startup without overriding the `strimzi.io/reconciled` annotation. If the operator errors during reconcile it will be clear that the currently reconciled version does not match the version that the operator is applying. A two stage commit mechanism here gives much more information regarding whether it started the reconciliation versus whether it finished successfully.
```